src/main.js

```diff
@@ -157,12 +166,66 @@ const createWindow = async () => {
       await shutdownPythonProcess();
     }
   });
+
+  // Setup handlers & defaults to manage downloads.
+  // Specifically, invest sample data downloads
+  let downloadDir;
+  let downloadLength;
+  const downloadQueue = [];
+  ipcMain.on('download-url', async (event, urlArray, directory) => {
+    logger.debug(`${urlArray}`);
+    downloadDir = directory;
+    downloadQueue.push(...urlArray);
+    downloadLength = downloadQueue.length;
+    mainWindow.webContents.send(
+      'download-status',
+      [(downloadLength - downloadQueue.length), downloadLength]
+    );
+    urlArray.forEach((url) => mainWindow.webContents.downloadURL(url));
+  });
+
+  mainWindow.webContents.session.on('will-download', (event, item) => {
```

Comment:
Where does this get called from? A quick `grep` didn't turn it up, but maybe I'll find it in some of the files I've yet to look through.

Reply:
Good question, 'will-download' is triggered by `mainWindow.webContents.downloadURL(url)`. 

This seems a bit convoluted, but I think it's the correct use of the API. The 'will-download' event needs to be fired from the `main` process because its emitter is on `mainWindow.webContents`. But the initial user-event of course comes from the `renderer`, so we have that 'download-url' listener as kind of the go-between channel between the renderer and the 'will-download' callback. 

I think this is another good reason to refactor `main` into smaller modules, so it can be easier to see which things are closely related.

Reply:
Ahhh interesting. Thanks for that. I'm so used to these emitters and listeners basically have the same corresponding names. Yea, I think the refactor of `main` will tidy things up.